File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq
def solution(jobs):
    answer = 0
    i = 0
    q = []
    start, now = -1, 0
    while i < len(jobs):
        for j in jobs:
            if start < j[0] <= now: # 얘는 시점을 가지고노는 변수라고 생각
                heapq.heappush(q, (j[1], j[0]))
        if q:
            take_time, now_time = heapq.heappop(q) # 소요시간 , 현재 시점
            start = now
            now = now + take_time
            answer = answer + (now - now_time)
            i += 1
            print(take_time, now ,now_time)
        else:
            now += 1 # 만약 처리할게 없다면 현재 처리시각을 1씩 늘리면서 검사를 시행한다
        
    return answer // len(jobs)

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        int answer = 0;
        int i = 0;
        int start = -1;
        int now = 0;
        
        PriorityQueue<int[]> heap = new PriorityQueue<>((o1,o2) -> Integer.compare(o1[0],o2[0]));

        while( i < jobs.length) {
            for (int [] j : jobs){
                if (start < j[0] && j[0] <= now){
                    heap.offer(new int[]{j[1],j[0]});
                }
            }
            if(!heap.isEmpty()){
                int [] current = heap.poll();
                int l = current[0];
                int s = current[1];
                start = now;
                now += l;
                answer += (now - s);
                i++;
            }else{
                now++;
            }
        }
        

        return answer / jobs.length;
    }
}